src/main/java/com/intuit/graphql/orchestrator/utils/FederationUtils.java

```diff
@@ -0,0 +1,120 @@
+        }
+
+        if(!fieldSet.startsWith("{")) {
+            fieldSet = StringUtils.join(StringUtils.SPACE, "{", fieldSet, "}");
+        }
+
+        //Throws InvalidSyntaxException if fieldSet is incorrect
+        Document fieldSetDocument = Parser.parse(fieldSet);
+
+        List<OperationDefinition> definitions = fieldSetDocument.getDefinitions().stream()
+                .map(OperationDefinition.class::cast).collect(Collectors.toList());
+
+        List<FieldDefinition> typeFieldDefinitions = getFieldDefinitions(typeDefinition);
+
+        for( final OperationDefinition definition : definitions) {
+            List<Field> fields = definition.getSelectionSet().getSelections().stream().map(Field.class::cast).collect(Collectors.toList());
+            for(Field field : fields) {
+                checkFieldReferenceRecursively(sourceGraph,typeDefinition.getName(), typeFieldDefinitions, field);
+            }
+        }
+    }
+
+    private static void checkFieldReferenceRecursively(XtextGraph sourceGraph, String typeName, List<FieldDefinition> declaredDefinitions, Field fieldToCheck) {
+        String fieldName = fieldToCheck.getName();
+        Optional<FieldDefinition> optionalFieldDefinition = declaredDefinitions.stream()
+                .filter(fieldDefinition -> fieldDefinition.getName().equals(fieldName))
+                .findFirst();
+
+        if(optionalFieldDefinition.isPresent()) {
+            if(CollectionUtils.isNotEmpty(fieldToCheck.getChildren())) {
+                FieldDefinition fieldDefinition = optionalFieldDefinition.get();
+                TypeDefinition childType = sourceGraph.getType(fieldDefinition.getNamedType());
+
+                //should only be 1 element, but put in loop just in case something changes later
+                for(SelectionSet childSelectionSet: fieldToCheck.getChildren().stream().map(SelectionSet.class::cast).collect(Collectors.toList())) {
+                    for(Field childField : getFieldsFromSelectionSet(childSelectionSet)) {
+                        checkFieldReferenceRecursively(sourceGraph, childType.getName(), getFieldDefinitions(childType), childField);
+                    }
+                }
+            }
+        } else {
+            throw new InvalidFieldSetReferenceException(fieldName, typeName);
+        }
+    }
+
+    private static List<Field> getFieldsFromSelectionSet(SelectionSet selectionSet) {
+        return selectionSet.getSelections().stream().map(Field.class::cast).collect(Collectors.toList());
+    }
+
+    public static boolean isFederationDirective(Directive directive) {
+        String directiveName = directive.getDefinition().getName();
+        return FED_TYPE_DIRECTIVES_NAMES_SET.contains(directiveName)
+            || FED_FIELD_DIRECTIVE_NAMES_SET.contains(directiveName);
+    }
+
+    public static boolean isBaseType(TypeDefinition entityDefinition) {
```

Comment:
isEntityBaseType

Reply:
Since we are not checking if the base type contains a key which would make it an entity, I would suggest not changing it to that.

Reply:
I should probably change the input name, so it isn't entityDefinition